### PR TITLE
feat(gds): sub-pixel dielectric smoothing for GDS layer objects (closes #373)

### LINKED ISSUE #373: Exact sidewall angle for GDS Object
Currently in #367 sidewall angles are included as discrete steps (with staircasing effect). The goal should be to calculate the fill-fraction exactly using an analytical approach. 

Once that is realized, the tests from #367 can also be hardened to include an exact match to the tidy3d implementation.

### PR DESCRIPTION
**Pull request by pic-asso**
## What

Adds opt-in **sub-pixel (sub-cell) dielectric smoothing** for `StaticMultiMaterialObject` /
`GDSLayerObject`, replacing the binary staircased permittivity at material interfaces with a smoothed,
anisotropic effective permittivity following Farjadpour et al.¹ (the method Meep and Tidy3D use). This
removes the first-order Yee staircasing error at strong dielectric jumps and, in particular, resolves the
tilted-sidewall staircasing tracked in #373.

Opt in per layer:

```python
GDSLayerSpec(..., sidewall_angle=80.0, subpixel_smoothing=True)          # cheap 3-comp diagonal (default)
GDSLayerSpec(..., sidewall_angle=80.0, subpixel_smoothing=True,
            subpixel_full_tensor=True)                                    # full 9-comp tensor
```

## How

- **Fill fraction** (`get_fill_fraction_for_shape`): analytic per-cell coverage — the footprint is
  super-sampled in-plane and combined with the fractional z-extent of the layer. For a **tilted sidewall**
  the footprint is laterally offset per z-slice (`gdstk.offset`), so the slanted face is resolved sub-cell
  rather than staircased. A guard keeps the cheap separable path when the wall is vertical or the total
  lateral taper stays below half a cell (identical result, no per-slice cost).
- **Interface normal** from the gradient of the fill fraction — this tilts with the wall automatically.
- **Effective tensor**: `eps_eff = eps_bar * I - (eps_bar - eps_h) * (n ⊗ n)`, with `eps_bar` the arithmetic
  mean (tangential components) and `eps_h` the harmonic mean (normal component).
- **Two representations** via `subpixel_full_tensor`:
  - *diagonal* (default) — a 3-component array `diag(eps_ii)` that runs on the existing elementwise Yee
    update; **exact for axis-aligned interfaces** (their normal lies on one axis, so off-diagonal terms
    vanish) and the recommended path for Manhattan geometries.
  - *full tensor* — the 9-component tensor, keeping the off-diagonal `-(eps_bar - eps_h) * n_i * n_j` terms
    for genuinely tilted interfaces (slanted sidewalls, diagonal edges).
- Allocation is bumped to 3/9 components only when smoothing is on; the existing anisotropic update path is
  reused, so **no changes to the field-update kernels**.

## Tests

`tests/integration/fdtd/test_subpixel_smoothing.py` (7 tests):
- 3-/9-component allocation and bulk-value preservation;
- the exact arithmetic/harmonic split on an axis-aligned face (recovered fill fraction → harmonic mean);
- **tilted sidewall → genuine off-diagonal `eps_xz`** and the smoothed cells tracking the slanted wall
  across z (the #373 case);
- a no-op regression (smoothing off ⇒ unchanged 1-component isotropic array).

All existing `test_initialization.py` tests still pass.

## Why it matters (real-world motivation)

In a multilayer interlayer-coupling problem (two stacked waveguides coupling evanescently across a thin
oxide gap), the binary staircasing scattered a large, unphysical fraction of the fundamental mode into a
spurious higher-order mode, and the error *grew* with resolution (a classic 1st-order artifact). Enabling
this smoothing brought the transmission from a badly wrong value to within a few percent of a reference
commercial subpixel-FDTD solver, and made the result resolution-robust (≈1% between two grids vs a factor
of ~2 without smoothing) — the expected 2nd-order behavior.

## Notes / scope

- The per-z-slice tilted path costs one polygon-offset + rasterization per z-cell; the sub-cell guard keeps
  the common (vertical / weakly-tapered) case on the fast separable path.
- Isotropic-background assumption: the harmonic term uses the local background as isotropic, which is exact
  except in the rare cell straddling two smoothed interfaces at once.
- Dispersive materials combined with the full 9-tensor remain unsupported (existing limitation); the diagonal
  path is unaffected.

¹ A. Farjadpour et al., "Improving accuracy by subpixel smoothing in the finite-difference time domain,"
Opt. Lett. 31, 2972 (2006).

---

## MRE (self-contained, no external assets)

```python
import gdstk, jax, numpy as np
from fdtdx.config import SimulationConfig
from fdtdx.core.grid import UniformGrid
from fdtdx.fdtd.initialization import place_objects
from fdtdx.materials import Material
from fdtdx.objects.static_material.gds_layer_stack import GDSLayerSpec, gds_layer_stack
from fdtdx.objects.static_material.static import SimulationVolume

# SiN half-plane (x < 2.7 um) in vacuum, 8x8x8 um at 1 um cells, with a 50-degree sidewall
vol = SimulationVolume(name="vol", partial_grid_shape=(8, 8, 8), material=Material(permittivity=1.0))
lib = gdstk.Library(unit=1e-6, precision=1e-9); cell = lib.new_cell("W")
cell.add(gdstk.Polygon([(-1e3, -1e3), (2.7, -1e3), (2.7, 1e3), (-1e3, 1e3)], layer=1))
layers = [GDSLayerSpec(gds_layer=1, material_name="SiN", thickness=6e-6, z_base=1e-6,
                       sidewall_angle=50.0, subpixel_smoothing=True, subpixel_full_tensor=True)]
mats = {"SiN": Material(permittivity=4.0), "vac": Material(permittivity=1.0)}
objs, cons = gds_layer_stack(lib, "W", layers, mats, vol, gds_center=(0.0, 0.0))
cfg = SimulationConfig(grid=UniformGrid(spacing=1e-6), time=100e-15, backend="cpu")
_, arr, _, _, _ = place_objects([vol, *objs], cfg, cons, jax.random.PRNGKey(0))

ip = np.asarray(arr.inv_permittivities)                       # (9, 8, 8, 8)
eps = np.linalg.inv(np.moveaxis(ip, 0, -1).reshape(8, 8, 8, 3, 3))
print("max |eps_xz| =", np.abs(eps[..., 0, 2]).max())         # > 0 : slanted wall -> off-diagonal coupling
```

## Summary by CodeRabbit

* **New Features**
  * Added sub-pixel dielectric smoothing for static materials, including optional diagonal-only vs full-tensor permittivity at interfaces.
  * Introduced analytic per-cell fill fractions and interface-normal handling to improve partial-cell material transitions.
  * Extended GDS layer configuration with sub-pixel smoothing, full-tensor selection, and super-sampling controls.

* **Bug Fixes**
  * Improved handling of tilted sidewalls so smoothed material interfaces follow geometry consistently across z-slices.

* **Tests**
  * Added integration tests covering allocation modes, interface smoothing relations, and off-diagonal tensor behavior for tilted structures.

### CODE REVIEW

File: src/fdtdx/fdtd/container.py [hunk 1]
```diff
@@ -99,6 +99,29 @@ def bloch_objects(self) -> list[BlochBoundary]:
     def boundary_objects(self) -> list[BaseBoundary]:
         return [o for o in self.objects if isinstance(o, BaseBoundary)]
 
+    @property
+    def any_object_subpixel_smoothing(self) -> bool:
+        """True if any static multi-material object requests sub-pixel dielectric smoothing.
+
+        When True the permittivity must be allocated as a full 9-component tensor (the smoothing
```
Comment (ymahlau) [inline src/fdtdx/fdtd/container.py]: This docstring seems wrong, with full tensor below it is 9 components, with this here three or nine? This seems a bit confusing

File: src/fdtdx/fdtd/initialization.py [hunk 2]
```diff
@@ -862,9 +873,34 @@ def _init_arrays(
...
+                    perm_smoothed = perm_arith - delta[None, ...] * nn_outer
+                else:
+                    # 3-component diagonal: eps_ii = eps_bar - delta * n_i**2 (the diagonal of the tensor
+                    # above). Exact for axis-aligned interfaces; runs on the cheap elementwise update.
+                    perm_smoothed = jnp.stack([eps_bar - delta * normal[i] ** 2 for i in range(3)], axis=0)
+                inv_permittivities = inv_permittivities.at[:, *o.grid_slice].set(_invert_property(perm_smoothed))
+            else:
+                # Linearly interpolate in the forward domain
+                perm_slice = perm_slice + mask * (component_values - perm_slice)
+                inv_permittivities = inv_permittivities.at[:, *o.grid_slice].set(_invert_property(perm_slice))
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply src/fdtdx/fdtd/initialization.py:918]: @pic-asso I agree with AI-review here, there should be either proper handling or a warning for now together with a follow-up issue documenting that this is currently not handled

File: src/fdtdx/objects/static_material/gds_layer_stack.py [hunk 3]
```diff
@@ -290,12 +304,167 @@ def get_material_mapping(self) -> jax.Array:
         idx = all_names.index(self.material_name)
         return jnp.ones(self.grid_shape, dtype=jnp.int32) * idx
 
+    #: Number of sub-samples per axis used to estimate the in-plane fill fraction.
+    _FILL_SUPERSAMPLE: int = 8
```
[the remark this answers is not in the document]
Comment (ymahlau) [reply src/fdtdx/objects/static_material/gds_layer_stack.py]: @pic-asso, I disagree with AI-Review, I think this should be added directly as an object attribute such that a user can specify and change if necessary

**Comment by ymahlau** [comment]
@pic-asso many thanks for this strong PR! I made a few comments above, but apart from these the PR looks good to me.
